Integer

desmonddak opened the pull request:
<!-- Please make sure you check out the contribution guidelines before submitting a pull request! -->

## Description & Motivation

Adds a very general booth encoding multiplier with important subcomponents:  generalized booth encoding (2,4,8,16 currently supported), Wallace tree column compression (delay-driven with heuristic fixed delays for column compressors), and reusing the parallel prefix adder components for the final adder.  

The multiplier base supports different widths of operands (so rectangular products) as well as signed and unsigned operands. Testing is quite exhaustive across signed, radix, width, skew (width difference).

Also, we added a multiply-accumulate which is simply introducing the third operand as part of the column compression.

## Related Issue(s)

None.

## Testing

Randomized testing to cover the wide space of radix, widths, skew, and signage.  Using a full multiply-accumulate tester to test both MAC and multiply by using a zero-term for the third operand to keep test code crisp.  Especially tricky has been compact sign extension (keeping the number of rows the same) across different widths and skews for signed and unsigned and this was tested exhaustively.

## Backwards-compatibility

> Is this a breaking change that will not be backwards-compatible? If yes, how so?

Changed existing multipliers to use an abstract Multiply module base class.

## Documentation

> Does the change require any updates to documentation? If so, where? Are they included?

Documentation updated with the multiplier classes.  Also updated the confapp generator to allow for web-based visualization.

mkorbel1 commented on test/arithmetic/multiplier_test.dart:
```diff
@@ -0,0 +1,221 @@
+    final signed = mod.signed;
+    for (var i = 0; i < iterations; i++) {
+      final bA = signed
+          ? Random().nextLogicValue(width: width).toBigInt().toSigned(width)
```
use a single `Random` object created in one place (maybe per-test?) with a fixed seed provided so that tests are reproducible

desmonddak replied:
done.

mkorbel1 commented on test/arithmetic/multiplier_test.dart:
```diff
@@ -0,0 +1,221 @@
+
+  MultiplyAccumulate curryUnsignedMultiplierAsMAC(Logic a, Logic b, Logic c) =>
+      MultiplyOnly(a, b, c, curryUnsignedMultiplier);
+
```
this is some really beautiful testing automation!

desmonddak replied:
Again, another area where I feel like I am getting the hang of Dart and can do this kind of thing much faster than I was able to in Chisel.  The only place were I felt Scala was a little cleaner was passing back values from control structures, other than that I have been more than pleasantly surprised how easily I could mix in functional style where it helps.

mkorbel1 reviewed (approved):
This is great!

mkorbel1 reviewed (changes requested):
whoops meant to request changes, not approve

desmonddak commented on lib/src/arithmetic/booth.dart:
```diff
@@ -0,0 +1,744 @@
+            6 => (extendedMultiplicand << 3) - (extendedMultiplicand << 1),
+            7 => (extendedMultiplicand << 3) - extendedMultiplicand,
+            8 => extendedMultiplicand << 3,
+            _ => extendedMultiplicand
```
I think this is a good idea.  I also have some other areas (such as when the width is too narrow to produce enough rows, there is a problem in that it happily emits an incorrect result).  E.g., width=3 I think is too narrow for radix=8.

desmonddak replied:
Added exception throw.  Still need to look at widths for multipliers to check and throw.

desmonddak commented on doc/components/multiplier.md:
```diff
@@ -1,21 +1,42 @@
 # Multiplier
 
-ROHD HCL provides a multiplier module to get the product from a list of logic. As of now, we have
+ROHD HCL provides an abstract [Multiplier] module which multiplies two
```
Ah.. I will go through and fix these.

desmonddak commented on lib/src/arithmetic/booth.dart:
```diff
@@ -0,0 +1,744 @@
+import 'package:rohd_hcl/src/utils.dart';
+
+/// Simplest version of bit string representation
+String bitString(LogicValue value) => value.toString(includeWidth: false);
```
Yes.  But I wasn't sure how to get that done and rolled out.  I did look into LogicValue itself, but I didn't know how to do extensions.  I will try this.  There are a couple of other functions I would like to model with extension and then if they are really useful, we could migrate into ROHD proper.  This helps a lot.

desmonddak commented on test/arithmetic/booth_test.dart:
```diff
@@ -0,0 +1,208 @@
+          ..insert(3, [(4, 6, 4), (8, 7, 5), (16, 10, 6)])
+          ..insert(4, [(8, 9, 6), (16, 9, 7)])
+          ..insert(5, [(16, 8, 8)])
+        //
```
Ah -- this is because it made it easier to comment out a row for testing.  A trick I use to bypass auto-formatting.

mkorbel1 replied:
ah very clever!  i guess can be removed though now that it's solid

desmonddak replied:
Done.

desmonddak commented on test/arithmetic/compressor_test.dart:
```diff
@@ -0,0 +1,116 @@
+// Copxorright (C) 2023 Intel Corporation
```
How do you catch these? :)

desmonddak commented on lib/src/arithmetic/adder.dart:
Updated all Copyright statements.

desmonddak commented on lib/src/arithmetic/booth.dart:
I realize now that I need to create a different package to hold these files for the internal book keeping.
I put them in rohd_hcl.dart as I needed them to behave as a package., but I need to put them somewhere else as a cluster of files that do these main things:  booth encoding, selection, pp generation, and compression.

desmonddak replied:
Created a multiplier_lib.dart to export the key symbols for multiplier building.

desmonddak commented on lib/src/arithmetic/booth.dart:
```diff
@@ -0,0 +1,744 @@
+  }
+
+  /// Accumulate the partial products and return as BigInt
+  BigInt evaluate() {
```
I found a really nice debugging ability was to be able to walk all structures and print out their current boolean values while in a test.  

I tried to create an extension, but I lose the ability to do a print() on the object (it won't allow me to define toString).  So I need to rethink.  This idea is definitely a test-time accessor to the classes, so an extension makes some sense.

mkorbel1 replied:
yeah it seems useful, just maybe not quite right where it lives (visible as public API).  you could always make a new function rather than override toString() in the extension

mkorbel1 replied:
I noticed that you moved it into an extension, but the extension is exported as part of the rohd_hcl lib, so it's still visible externally

desmonddak commented on test/arithmetic/compressor_test.dart:
```diff
@@ -0,0 +1,116 @@
+              'vs expected $product\n')
+          ..write(pp);
+      }
+      expect(pp.evaluate(), equals(product));
```
This is really helpful.  I was always putting prints before the failure because expect gave just the mismatch and then you were not really in a debug state where you could see values and print.  Now I can print out the failing partial product matrix and more quickly debug, yet not have print statements to comment and uncomment for debug.

desmonddak commented on lib/src/arithmetic/adder.dart:
```diff
@@ -21,9 +21,28 @@ abstract class Adder extends Module {
+  /// The addition results [sum] including carry bit
   Logic get sum => output('sum');
 
+  /// Implementation needs to provide a method for calculating the full sum
```
Done for existing adders and base computes the sum.

desmonddak replied:
Yet I now see a fundamental problem.  The base class constructor is called before the derived class.  So if the base class asks for CalculateOut()  to build the sum, that is a call to the derived class which has not yet built what it needs to compute the sum.  I have to create a prefix tree in the derived class  for example before I can assign out.  The alternative is I have to now have a Logic to hold the out to break this.

desmonddak replied:
My attempts on this all fail.  There is a circularity formed that makes it difficult to decompose the logic into base class and derived class when you look at the order of construction.  late initialization, getters, and and the addOutput() routine need a good design pattern defined for this and I'm just not figuring out the right pattern.  I was using a pattern of inputs only because the problem seems to be on the output side.

mkorbel1 replied:
Perhaps the child classes should be defining their output signal as a `final` variable (prior to call to `super`), then assigning the computed value into that output signal?

desmonddak replied:
I think this was the resolution:  use a Logic in the child class to break the cycle and pass back the value.
I think the adder base class is now a reasonable API.

desmonddak commented on lib/src/arithmetic/multiplier.dart:
```diff
@@ -0,0 +1,191 @@
+        PartialProductGenerator(a, b, RadixEncoder(radix), signed: signed);
+    // ignore: cascade_invocations
+    pp.signExtendCompact();
+    final compressor = ColumnCompressor(pp);
```
Cleanup needed.  I would test different sign extension routines, which is easier without cascade.

desmonddak commented on test/arithmetic/adder_test.dart:
```diff
@@ -0,0 +1,224 @@
+        final bJ = BigInt.from(j).toSigned(n);
+        final BigInt smaller;
+        final BigInt bigger;
+        // When equal, we want the negative one first to produce 1 1...1
```
I shouldn't swap, just continue without testing to cut the testing space in half.  The Adder requires the first operand be bigger if negative to get the sign right.  I'm thinking about how to handle this in a more general case to avoid a comparator when instantiated.

desmonddak replied:
I fixed to just continue when not bigger.

desmonddak replied:
constructor flag now to avoid a comparator internally if you've sorted the inputs externally.
One clumsiness is that if you have an equal magnitude positive and negative, the negative needs to come first as 'a'.

desmonddak commented on test/arithmetic/booth_test.dart:
```diff
@@ -0,0 +1,208 @@
+import 'package:rohd_hcl/src/utils.dart';
+import 'package:test/test.dart';
+
+enum SignExtension { brute, stop, compact, compactRect }
```
A challenge for debug is to print before and after each step when needed.  I guess I could pass in a debug setting.

This brings up my testing approach:  for a singleton case, I can wire up Logics already set with values at the input and then when I print the structures (like a matrix for the partial products), I can see their actual values and figure out what went wrong.  This is what my 'evaluate' and 'toString' methods do:  they are meant for live running of values during construction so we can see the bug with an assigned values.  In other comments, you talk about moving these to an extension.  That may complicate things because INSIDE the construction of the hardware I may want to print intermediate steps (like pre-sign extension and post sign extension, or after compression, to see how bits got mangled).  

So if I do both:  create a flag for the type of sign extension so that it is done inside the constructor AND move all evaluate routines outside the class, I cannot use this debug method (or I don't see an easy way).

desmonddak replied:
This is now done.  signExtension option.

desmonddak commented on test/arithmetic/booth_test.dart:
```diff
@@ -0,0 +1,208 @@
+    // expect(pp.evaluate(signed: true), equals(product));
+  });
+
+  // TODO(dakdesmond): Figure out minimum widths!
```
Yes.  I figured out the formula for minimum multiplicand and minimum multiplier widths as a function of radix.

desmonddak commented on lib/src/arithmetic/compressor.dart:
```diff
@@ -0,0 +1,333 @@
+//  We need a way to build the CompressionTerm without actual values
+//    e.g., there needs to be a way to do the reductions with 'X' values
+  /// Evaluate the logic value of a given CompressTerm
+  LogicValue evaluateTerm(CompressTerm term) {
```
Great suggestion.  

Again, I need to figure out how to express the debug API of my evaluate / print routines that are in use DURING simulation versus other constructive APIs.  They are hyper-useful (I can show you dozens of pages of these debug situations to figure out all the corner cases).  They would be useful if we find a bug and I can setup that particular set of values and do the evaluates.

desmonddak replied:
Refactored  as you recommended.

desmonddak commented on lib/src/arithmetic/compressor.dart:
```diff
@@ -0,0 +1,333 @@
+        ts.write('\t${bitString(rowBits.swizzle().zeroExtend(width))} ($val)');
+        if (row == rows - 1) {
+          ts.write(' Total=${accum.toSigned(width)}\n');
+          stdout.write(ts);
```
Fixed.  I also put in a demonstration of evaluate routines in the _test.

desmonddak commented on lib/src/arithmetic/sign_magnitude_adder.dart:
```diff
@@ -0,0 +1,69 @@
+
+  /// [SignMagnitudeAdder] constructor with an unsigned adder functor
+  SignMagnitudeAdder(Logic as, super.a, Logic bs, super.b,
+      Adder Function(Logic, Logic) adderGen)
```
Adders all behave as twos-complement (they don't have to know the Logic is signed).  in order to get it to behave as a magnitude only adder with an external sign, I have this Module.

desmonddak commented on lib/src/arithmetic/parallel_prefix_operations.dart:
```diff
@@ -166,12 +167,28 @@ class ParallelPrefixOrScan extends Module {
 }
 
+/// Priority Finder based on ParallelPrefix tree
+class ParallelPrefixPriorityFinder extends Module {
```
I would need to understand how the bit input to the arbiter really works to create a prefix tree to select.

desmonddak commented on lib/src/arithmetic/compressor.dart:
```diff
@@ -0,0 +1,333 @@
+  }
+
+  /// Create a sum Term
+  factory CompressTerm.sumTerm(List<CompressTerm> args, int row, int col) {
```
Refactored this completely to eliminate factory constructors and use proper constructor arguments for not only the input, but the computed output logic.

desmonddak commented on lib/src/arithmetic/booth.dart:
```diff
@@ -0,0 +1,744 @@
+  Logic select(int col, RadixEncode encode) =>
+      _select(getMultiples(col), encode);
+}
+
```
This is broken up into three files.  Also, I use a multiplier_lib.dart file to export their symbols.

desmonddak commented on lib/src/arithmetic/compressor.dart:
```diff
@@ -0,0 +1,333 @@
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/src/arithmetic/booth.dart';
+
+// TODO(desmonddak): Logic and LogicValue majority() functions
```
Done.  I put a draft form of LogicValue:majority() in as well.

desmonddak commented on lib/src/arithmetic/compressor.dart:
```diff
@@ -0,0 +1,333 @@
+// TODO(desmonddak): Logic and LogicValue majority() functions
+
+/// Base class for column compressor function
+class Compressor extends Module {
```
Done.

desmonddak commented on lib/src/arithmetic/compressor.dart:
```diff
@@ -0,0 +1,333 @@
+  }
+}
+
+// ignore: public_member_api_docs
```
Cleaned up.

desmonddak commented on lib/src/arithmetic/compressor.dart:
```diff
@@ -0,0 +1,333 @@
+          }
+          final t = CompressTerm.sumTerm(inputs, 0, col);
+          t.logic <= compressor.sum;
+          // assert(t.logic.value == evaluateTerm(t),
```
done.

desmonddak commented on lib/src/arithmetic/multiplier.dart:
```diff
@@ -0,0 +1,191 @@
+  /// Take input [a] and input [b], compute their
+  /// product, add input [c] to produce the [accumulate] result.
+  MultiplyAccumulate(Logic a, Logic b, Logic c,
+      {bool signed = false, super.name}) {
```
Done.

desmonddak commented on lib/src/arithmetic/multiplier.dart:
```diff
@@ -0,0 +1,191 @@
+
+    final pp =
+        PartialProductGenerator(a, b, RadixEncoder(radix), signed: signed);
+    // ignore: cascade_invocations
```
Folding sign extension into the constructor eliminated this debugging approach.

desmonddak commented on lib/src/arithmetic/multiplier.dart:
```diff
@@ -0,0 +1,191 @@
+    final pp =
+        PartialProductGenerator(a, b, RadixEncoder(radix), signed: signed);
+    // ignore: cascade_invocations
+    pp.signExtendCompact();
```
I now did this.  One option is 'none' to allow for debug of the matrix before sign extension.

desmonddak commented on lib/src/arithmetic/multiplier.dart:
```diff
@@ -0,0 +1,191 @@
+    // ignore: cascade_invocations
+    pp.signExtendCompact();
+
+    // Evaluate works only because the compressed rows have the same shape
```
I moved closer to where it is relevant.

desmonddak commented on lib/src/arithmetic/parallel_prefix_operations.dart:
```diff
@@ -184,32 +201,46 @@ class ParallelPrefixPriorityEncoder extends Module {
-  /// Output [out] the arithmetic sum of the two Logic inputs
-  Logic get out => output('out');
+class ParallelPrefixAdder extends Adder {
+  late final Logic _out;
```
Fixed.

desmonddak commented on lib/src/arithmetic/sign_magnitude_adder.dart:
```diff
@@ -0,0 +1,69 @@
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// An Adder which performs one's complement arithmetic using an unsigned
```
Fixed.

desmonddak commented on test/arithmetic/booth_test.dart:
```diff
@@ -0,0 +1,208 @@
+      final Y = pp.signed ? j.toSigned(widthY) : j.toUnsigned(widthY);
+      pp.multiplicand.put(X);
+      pp.multiplier.put(Y);
+      // print('$X * $Y');
```
The 'reason' field really helped.

desmonddak commented on test/arithmetic/booth_test.dart:
```diff
@@ -0,0 +1,208 @@
+      // ignore: cascade_invocations
+      // stdout.write(pp);
+    }
+    // expect(pp.evaluate(signed: true), equals(product));
```
expect is live now.

desmonddak commented on test/arithmetic/booth_test.dart:
```diff
@@ -0,0 +1,208 @@
+          for (var skew = -3; skew < shift * 2; skew++) {
+            // Only some sign extension routines have rectangular support
+            for (final signExtension in [
+              // SignExtension.brute,
```
Commented how I am doing this.

desmonddak commented on test/arithmetic/booth_test.dart:
```diff
@@ -0,0 +1,208 @@
+              case SignExtension.compactRect:
+                pp.signExtendCompactRect();
+            }
+            // testPartialProductExhaustive(pp);
```
Removed comment.

desmonddak commented on test/arithmetic/booth_test.dart:
```diff
@@ -0,0 +1,208 @@
+
+    // These collide with the normal q extension bits
+    // These are unsigned tests
+    // ignore: cascade_invocations
```
fixed

desmonddak commented on lib/src/arithmetic/booth.dart:
```diff
@@ -0,0 +1,744 @@
+  }
+
+  /// Sign extend the PP array using stop bits without adding a row
+  /// This routine works with different widths of multiplicand/multiplier
```
I added a reference for the base implementation.  I think I have publishable innovation on top which is to make it work for

1. Widths that are not multiples of the shift of the radix (e.g. shift of radix 8 is 3).  This was the first problem I discovered in the publication that they had conveniently chosen w=12 because it was a multiple of the shift and didn't work at all widths.
2. rectangular multipliers where the multiplier and multiplicand are different widths (I see that as a potential in ML dot-products -- perhaps a convenience, but perhaps that flexibility will become quite useful.
3. signed vs unsigned: the paper did not address, and it turns out to be interestingly tricky. I found that a Stanford thesis appendix which made it clear that you have to extend the multiplier when signed, not obvious at all!

desmonddak commented on lib/src/arithmetic/booth.dart:
```diff
@@ -0,0 +1,744 @@
+  int get shift => selector.shift;
+
+  /// The actual shift in each row
+  final rowShift = <int>[];
```
Sign extension routines modify the rowShift and partialProduct arrays to tuck in the signs.  Not really outside the class, but there is the need to extend the partialproduct and add a row (impacting both).

desmonddak commented:
I think I am done with updates.
I added a couple of proposed extensions to LogicValue because one I am now using (firstOne()) in printing during evaluation.
There are a couple of open issues to discuss:

1. A template for building input/output in a base Module and then deriving from that (the computeSum() issue).  I think we should create a template in the documentation as this is a starting point for adding a component and it can be a bit confusing where to use  `this.<input>` and `late final` and `get`.  A template and a few simple examples would go a long way.
2. The yosys bug and need for sv2v for the schematic (I added the multiplier to the web configurator for example)
3. My use of during-simulation `evaluate `code and how to structure that as an extension in test area only.
4. The extensions I added -- are these ok for now
5. A fair amount of restructured code based on your suggestions (folding in the sign extension into the constructor, getting rid of factory methods, etc.

desmonddak commented:
I have one more set of changes I want to make: generalizing the full multiply tests.

mkorbel1 commented on lib/src/arithmetic/multiplier.dart:
```diff
@@ -0,0 +1,191 @@
+  Multiplier(Logic a, Logic b, {bool signed = false, super.name}) {
+    _signed = signed;
+    this.a = addInput('a', a, width: a.width);
+    this.b = addInput('b', b, width: b.width);
```
Sure, we can work through it

mkorbel1 commented on lib/src/utils.dart:
```diff
@@ -1,10 +1,53 @@
+    var shiftedValue = this;
+    var result = 0;
+    while (shiftedValue != zero) {
+      result++;
```
shouldn't this be `result += (shiftedValue & 1)`?  otherwise, wouldn't `01000000` show as majority since it's just counting bits (0 or 1) to the right of the uppermost 1?

mkorbel1 commented on lib/src/utils.dart:
```diff
@@ -1,10 +1,53 @@
+    while (shiftedValue[0] != LogicValue.one) {
+      result++;
+      if (result == width) {
+        result = -1;
```
maybe this is a good opportunity to return `int?` and return null if none is found?

mkorbel1 reviewed (commented):
new changes look good!

desmonddak commented:
I'm working on documentation now... 
I'm not sure if this PR has gone on too long and we should reopen to clear out too much history and small changes/bugfixes.

desmonddak commented:
A few areas for focusing comments:
- documentation that I added -- especially how much I should reference args and describe calling (I provide examples).
- multiplier_lib.dart -- I had to pull this into rohd_hcl.dart anyway to get documentation produced, so we may need to remove this file and merge into rohd_hcl.dart for simplicity.
- tests:  I run regressions but can trim them down more.  I like this idea of having singleton tests and then the regression loops as when a bug surfaces it is easier to reconstruct using a singleton.
- an idea I had:  it would be nice to construct small examples that get run in the test area but get `included` in the .md files -- that way we don't end up with documentation that is incorrect in terms of calling conventions.
-  I'm tempted to create another PR to remove all this clumsy history.

mkorbel1 commented:
> I'm not sure if this PR has gone on too long and we should reopen to clear out too much history and small changes/bugfixes.

@desmonddak doesn't matter, this will all get squashed into one nice pretty commit when it merges :)

mkorbel1 commented on lib/rohd_hcl.dart:
```diff
@@ -1,10 +1,15 @@
 
-export 'src/adder.dart';
 export 'src/arbiters/arbiters.dart';
+export 'src/arithmetic/adder.dart';
```
We can add an "arithmetic.dart" that exports everything that we want to from the arithmetic directory instead of adding all these at the top-level here.

When you do so, review which things we want to export visibly to anyone and which should remain internal to rohd-hcl's implementation.  Anything not exported won't be automatically included when people import rohd_hcl

desmonddak replied:
I organized things a bit better to allow us to only export the main classes.  I created an arithmetic.dart file to export.

mkorbel1 commented on lib/src/arithmetic/sign_magnitude_adder.dart:
```diff
@@ -0,0 +1,62 @@
+
+    final bLarger = a.lt(b) | (a.eq(b) & bSign.gt(aSign));
+
+    final computeSign = mux(largestMagnitudeFirst ? Const(1) : Const(0), aSign,
```
You can use the dart ternary operator directly for generation-time stuff, and `mux` for hardware.

I think this could just be 
```dart
    final computeSign = largestMagnitudeFirst ? aSign : mux(bLarger, bSign, aSign);
```

You could also maybe avoid creation of the `bLarger` logic by using more of an `if`/`else` block here instead, even though it would synthesize ok you might end up with extra SV and extra simulation cost (small).

desmonddak replied:
I fixed this.  There are cases where it is a little harder:  see my new onesComplementAdder which takes either a Logic subtractIn or a boolean subtract.  It's hard to avoid converting the boolean to a signal to help.  But this is an interesting case:  if we want either a runtime configuration option or a compile time option that do the same thing.

mkorbel1 commented on lib/src/arithmetic/sign_magnitude_adder.dart:
```diff
@@ -0,0 +1,62 @@
+    bSign = addInput('bSign', bs);
+    final sign = addOutput('sign');
+
+    final bLarger = a.lt(b) | (a.eq(b) & bSign.gt(aSign));
```
Does this properly consider the case of `a != b` and the signs are different?

desmonddak replied:
sign_magnitude tests are exhaustive.

mkorbel1 commented on lib/src/arithmetic/sign_magnitude_adder.dart:
```diff
@@ -0,0 +1,62 @@
+  bool largestMagnitudeFirst;
+
+  /// [SignMagnitudeAdder] constructor with an unsigned adder functor
+  SignMagnitudeAdder(Logic as, super.a, Logic bs, super.b,
```
why not have these as `aSign` and `bSign` as well?

mkorbel1 commented on lib/src/arithmetic/sign_magnitude_adder.dart:
```diff
@@ -0,0 +1,62 @@
+      : super(
+            name: 'Ones Complement Adder: '
+                '${adderGen.call(Logic(), Logic()).name}') {
+    aSign = addInput('aSign', as);
```
you can re-assign variables from the constructor instead of making these `@protected late final` variables, unless there's some reason you want them exposed/inherited?

desmonddak replied:
Somehow I got stuck with a combinational loop, (input already assigned), but I managed to fix by using straight Logic in the class.

mkorbel1 commented on lib/src/arithmetic/parallel_prefix_operations.dart:
```diff
@@ -161,54 +161,79 @@ class ParallelPrefixOrScan extends Module {
-      ParallelPrefix Function(List<Logic>, Logic Function(Logic, Logic))
-          ppGen) {
+  ParallelPrefixOrScan(Logic inp,
+      ParallelPrefix Function(List<Logic>, Logic Function(Logic, Logic)) ppGen)
```
Should we add defaults on these too while we're at it?  Just to make them easier to use?

desmonddak replied:
Done

mkorbel1 commented on lib/src/arithmetic/partial_product_generator.dart:
```diff
@@ -0,0 +1,561 @@
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// Methods for sign extending the [PartialProductGenerator]
+enum SignExtension {
```
Upon rereading this, perhaps this is an opportunity to use a base class with sub-classes that extend it rather than an enum with a case statement?

desmonddak replied:
I created a set of subclasses, kept one in the main file and moved the others to a test file.   I use the simpler ones for debug to make sure I don't have a sign extension error that is blocking, or when doing an enhancement (like accumulate).

mkorbel1 commented on test/arithmetic/multiplier_encoder_test.dart:
```diff
@@ -0,0 +1,238 @@
+  test('Rectangle Q collision tests,', () async {
+    // These collide with the normal q extension bits
+    // These are unsigned tests
+    final alignTest = <List<(int, int, int)>>[]
```
Why use `insert` instead of just building the list as a literal?

desmonddak commented on doc/components/multiplier_components.md:
```diff
@@ -0,0 +1,202 @@
+
+Once you have a partial product matrix, you would like to add up the addends.  Traditionally this is done using compression trees which instantate 2:1 and 3:2 column compressors (or carry-save adders) to reduce the matrix to two addends.  The final two addends are often added with an efficient final adder.
+
+Our [`ColumnCompressor`] class uses a delay-driven approach to efficiently compress the rows of the partial product matrix.  Its only arguement is a [`PartialProductGenerator`] , and it creates a list of ColumnQueues containing the final two addends stored by column after compression. An extractRow() routine can be used to extract the columns.  [`ColumnCompressor`] currently has an extension [`Ev ...
```
Using that now, it really helps.

desmonddak commented on lib/src/arithmetic/multiplier.dart:
```diff
@@ -0,0 +1,170 @@
+  /// Construct a compression tree integer multipler with
+  ///   a given radix and final adder functor
+  CompressionTreeMultiplier(super.a, super.b, int radix,
+      ParallelPrefix Function(List<Logic>, Logic Function(Logic, Logic)) ppTree,
```
I put in defaults where I could.  should be a lot easier now.

mkorbel1 commented on lib/rohd_hcl.dart:
```diff
@@ -1,10 +1,10 @@
 
-export 'src/adder.dart';
 export 'src/arbiters/arbiters.dart';
+export 'src/arithmetic/adder.dart';
```
Why is adder.dart not part of the arithmetic.dart exports?

desmonddak replied:
I missed this

mkorbel1 commented on lib/src/arithmetic/addend_compressor.dart:
```diff
@@ -0,0 +1,253 @@
+}
+
+/// A column of partial product terms
+typedef ColumnQueue = PriorityQueue<_CompressTerm>;
```
Should this be private? `_ColumnQueue`?

desmonddak replied:
I think I need to factor out a module out of this to create a real API which is in: partialproduct out: LogicArray [2].
During testing, I didn't do this.  This will take me a bit to refactor (maybe not too bad).

desmonddak replied:
I've done the refactoring to hide ColumnCompression classes and wrap with an AddendCompression class.

mkorbel1 commented on lib/src/arithmetic/parallel_prefix_operations.dart:
```diff
@@ -161,54 +161,83 @@ class ParallelPrefixOrScan extends Module {
+  ParallelPrefixAdder(super.a, super.b,
+      {ParallelPrefix Function(List<Logic>, Logic Function(Logic, Logic))
+          ppGen = KoggeStone.new})
+      // : _out = Logic(width: a.width),
```
delete old code?

mkorbel1 commented on lib/src/arithmetic/partial_product_generator.dart:
```diff
@@ -0,0 +1,260 @@
+
+  /// Perform sign extension (defined in child classes)
+  @protected
+  void signExtend() {}
```
Shoudl this be an `abstract` class with no definition for this function?  Maybe the "none" version should be a baseclass that overrides this function and does nothing?

desmonddak replied:
I wonder if I should call this class PartialProductGeneratorBase.

Then my default PartialProductGenerator can really be the PartialProductGenerator_WithCompactRectSIgnExtension.  I don't want to lose the other cases because when we are testing a new kind of arithmetic, we may have to drop back to, say, BruteForce, just to get something working and then fix whatever is broken in the fancier, optimized sign extension.

mkorbel1 replied:
Sure, that sounds good

mkorbel1 commented on doc/components/multiplier_components.md:
````diff
@@ -0,0 +1,202 @@
+```dart
+    final pp =
+        PartialProductGenerator(a, b, RadixEncoder(radix), signed: true, 
+        signExtension: SignExtension.compactRect);
````
this example needs updating? no more enum in the constructor?

mkorbel1 commented on lib/src/arithmetic/partial_product_generator.dart:
```diff
@@ -0,0 +1,260 @@
+    }
+  }
+
+  // /// Sign extend the PP array using stop bits without adding a row
```
weird double-commenting?

mkorbel1 commented on test/arithmetic/adder_test.dart:
```diff
@@ -0,0 +1,339 @@
+        // When equal, we want the negative one first to produce 1 1...1
+        if (sortOperands &
+            ((bI.abs() < bJ.abs()) || (bI.abs() == bJ.abs() && (bI > bJ)))) {
+          continue;
```
Is this not filtering it such that we don't test the ability for the hardware to handle sorting the operands itself?

desmonddak replied:
If sortOperands is false, then the hardware is doing the sorting.  If sortOperands is true, we need to continue past this case as the operands are in bad order.

desmonddak replied:
I renamed to operandsArePreSorted

mkorbel1 replied:
ahh i see, thanks!

mkorbel1 reviewed (approved):
Looks good!
